devito/finite_difference.py

```diff
@@ -221,3 +224,196 @@ def second_cross_derivative(function, dims, order):
+    # Check number of dimension and which dimension to set offset to
+    if hasattr(func, 'space_dimensions'):
+        spc_dims = func.space_dimensions
+    else:
+        spc_dims = [f.space_dimensions for f in func.args
+                    if hasattr(f, 'space_dimensions')][0]
+    offsets = lambda x, y: dict([(d, 0) if d != x else (x, y)
+                                 for d in spc_dims])
+    # Computes fd offsets and cofficients
+    indices = [(dim + i * dim.spacing) for i in range(-fd_order, fd_order + 1)]
+    coeffs = finite_diff_weights(deriv_order, indices, dim)[-1][-1]
+    coeffs = [c.subs({dim: 0}) for c in coeffs[::-1]]
+    indices = [offsets(dim, i.subs({dim: 0})) for i in indices]
+    # Build fd List
+    if isinstance(func, sparse_fd_list):
+        fd = [(w*f[0], sum_dict(pos, f[1])) for w, pos in zip(coeffs, indices)
+              for f in func]
+    else:
+        fd = sparse_fd_list([(w*func, pos) for w, pos in zip(coeffs, indices)])
+    return fd.simplify()
+
+
+def sparse_cross_derivative(func, dims, fd_order):
+    """
+    Derives cross derivative for a product of given sparse functions. This does not
+    return a finite-difference expression but a list of (weight, offset) to construct
+    the interpolation at the finite-difference locations.
+
+    :param dims: 2-tuple of symbols defining the dimension wrt. which
+       to differentiate, eg. `x`, `y`, `z`.
+    :returns: The cross derivative
+    """
+    # Check number of dimension and which dimension to set offset to
+    if hasattr(func, 'space_dimensions'):
+        spc_dims = func.space_dimensions
+    else:
+        spc_dims = [f.space_dimensions for f in func.args
+                    if hasattr(f, 'space_dimensions')][0]
+
+    offsets = lambda x, y: dict([(d, 0) for d in spc_dims if d not in x] +
+                                [(xx, yy) for xx, yy in zip(x, y)])
+    # First dimension
+    indices1 = [(dims[0] + i * dims[0].spacing/2) for i in range(-fd_order, fd_order + 1)]
+    coeffs1 = finite_diff_weights(1, indices1, dims[0])[-1][-1]
+    coeffs1 = [c.subs({dims[0]: 0}) for c in coeffs1[::-1]]
+    indices1 = [i.subs({dims[0]: 0}) for i in indices1]
+    # Second dimension
+    indices2 = [(dims[1] + i * dims[1].spacing/2) for i in range(-fd_order, fd_order + 1)]
+    coeffs2 = finite_diff_weights(1, indices2, dims[1])[-1][-1]
+    coeffs2 = [c.subs({dims[1]: 0}) for c in coeffs2[::-1]]
+    indices2 = [i.subs({dims[1]: 0}) for i in indices2]
+    if isinstance(func, sparse_fd_list):
+        cross = [(w1*w2*f[0], sum_dict(f[1], offsets(dims, (h1, h2))))
+                 for (w2, h2) in zip(coeffs2, indices2)
+                 for (w1, h1) in zip(coeffs1, indices1) for f in func]
+    else:
+        cross = [[(w1*w2*func, offsets(dims, (h1, h2)))
+                 for (w2, h2) in zip(coeffs2, indices2)]
+                 for (w1, h1) in zip(coeffs1, indices1)]
+    return sparse_fd_list(cross).simplify()
+
+
+class sparse_fd_list(list):
```

Comment:
Using a ``namedtuple`` for the entries of list might be helpful. It would allow us to drop all that explicit indexing in the implementation below. Nitpicking, though; I'll leave this up to u